hazelcast/src/main/java/com/hazelcast/internal/config/override/ConfigProvider.java

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hazelcast.internal.config.override;
+
+import java.util.Map;
+
+/**
+ * An interface representing a generic configuration source.
+ */
+interface ConfigProvider {
```

Comment:
Minor: maybe you can join the "parser" and "provider" as right now the providers are rather simple. Right now, it looks like you might use different parsers with a single provider but it's not really needed. The provider implementation could both fetch the properties and parse them.

Reply:
We'd lose testability. Parsers work on arbitrary sources of data while providers supply parsers with data from `System.getEnv/getProperties`. Thanks to the separation, I'm able to test the parsing logic without touching `System.getEnv/getProperties`